plugins/audio-s5gen2/fu-audio-s5gen2-ble-device.c

```diff
@@ -0,0 +1,615 @@
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "GAIA protocol %u.%u is not supported",
+			    self->api_version.major,
+			    self->api_version.minor);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/* read the list of supported features from device */
+static gboolean
+fu_qc_s5gen2_ble_device_get_features(FuQcS5gen2BleDevice *self, gboolean next, GError **error)
+{
+	guint8 buf[FU_QC_S5GEN2_BLE_DEVICE_BUFFER_SZ] = {0};
+	gsize read_len;
+	g_autoptr(GByteArray) req = fu_struct_qc_gaia_v3_supported_features_req_new();
+	g_autoptr(GByteArray) resp = NULL;
+
+	fu_struct_qc_gaia_v3_supported_features_req_set_vendor_id(req, self->vid_v3);
+
+	fu_struct_qc_gaia_v3_supported_features_req_set_command(
+	    req,
+	    (next == FALSE) ? FU_QC_GAIA_V3_CMD_GET_SUPPORTED_FEATURES_REQ
+			    : FU_QC_GAIA_V3_CMD_GET_SUPPORTED_FEATURES_NEXT_REQ);
+
+	if (!fu_qc_s5gen2_ble_device_send(self, req->data, req->len, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_recv(self, buf, sizeof(buf), &read_len, error))
+		return FALSE;
+
+	resp = fu_struct_qc_gaia_v3_supported_features_parse(buf, read_len, 0, error);
+	if (resp == NULL)
+		return FALSE;
+
+	/* must be odd: header 5B + feature pairs */
+	if (!(read_len & 0x01)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "got incorrect features list");
+		return FALSE;
+	}
+
+	/* parse feature:version pairs */
+	for (gsize i = FU_STRUCT_QC_GAIA_V3_SUPPORTED_FEATURES_SIZE; i < read_len; i += 2) {
+		switch (buf[i]) {
+		case FU_QC_GAIA_V3_FEATURES_CORE:
+			self->feature.core = buf[i + 1];
+			g_debug("Core feature version: %u", self->feature.core);
+			break;
+		case FU_QC_GAIA_V3_FEATURES_DFU:
+			self->feature.dfu = buf[i + 1];
+			g_debug("DFU feature version: %u", self->feature.dfu);
+			break;
+		default:
+			break;
+		}
+	}
+
+	/* request the rest of the list */
+	if (fu_struct_qc_gaia_v3_supported_features_get_more_features(resp) == FU_QC_MORE_MORE)
+		return fu_qc_s5gen2_ble_device_get_features(self, TRUE, error);
+
+	return TRUE;
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_get_serial(FuQcS5gen2BleDevice *self, GError **error)
+{
+	guint8 buf[FU_QC_S5GEN2_BLE_DEVICE_BUFFER_SZ] = {0};
+	gsize read_len;
+	g_autoptr(GByteArray) req = fu_struct_qc_gaia_v3_serial_req_new();
+	g_autoptr(GByteArray) validate = NULL;
+	g_autofree gchar *serial = NULL;
+
+	fu_struct_qc_gaia_v3_serial_req_set_vendor_id(req, self->vid_v3);
+
+	if (!fu_qc_s5gen2_ble_device_send(self, req->data, req->len, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_recv(self, buf, sizeof(buf), &read_len, error))
+		return FALSE;
+
+	/* Check if response is valid */
+	validate =
+	    fu_struct_qc_gaia_v3_serial_parse(buf, FU_STRUCT_QC_GAIA_V3_SERIAL_SIZE, 0, error);
+	if (validate == NULL)
+		return FALSE;
+
+	serial = g_strndup((gchar *)(buf + FU_STRUCT_QC_GAIA_V3_SERIAL_SIZE),
+			   read_len - FU_STRUCT_QC_GAIA_V3_SERIAL_SIZE);
```

Comment:
this scares me a bit, can we use fu_strsafe instead pls.

Reply:
as far as I understood, NULL should be safe for `fu_device_set_serial()`, correct?

Reply:
Yup, that's fine. Just not `\x01\x02\x03\x04` :)